File: Prosary.Tests/FranciscanCrownEngineTests.cs
using Prosary.Models;
using Prosary.Services;
using Xunit;

namespace Prosary.Tests;

/// <summary>Mirrors iOS's FranciscanCrownEngineTests.swift / Android's FranciscanCrownEngineTest.kt.</summary>
public class FranciscanCrownEngineTests
{
    private readonly PrayerEngine _engine = new(new LiturgicalCalendarService());

    private static IReadOnlyList<RosaryStep> BuildSteps(PrayerEngine engine, string? languageCode) =>
        engine.BuildSteps(new Prayer { Kind = PrayerKind.FranciscanCrown, LanguageCode = languageCode ?? LanguageCatalog.DefaultSentinel });

    [Fact]
    public void BuildSteps_SevenDecadesOfTenHailMarys()
    {
        var steps = BuildSteps(_engine, "en");
        var decadeIndices = steps.Where(s => s.DecadeIndex.HasValue).Select(s => s.DecadeIndex!.Value).Distinct();
        Assert.Equal(Enumerable.Range(0, 7), decadeIndices.OrderBy(i => i));

        for (var d = 0; d < 7; d++)
        {
            var hailMarysInDecade = steps.Count(s => s.DecadeIndex == d && s.HailMaryIndexInDecade.HasValue);
            Assert.Equal(10, hailMarysInDecade);
        }
    }

    [Fact]
    public void BuildSteps_NoStepHasAMystery()
    {
        // Franciscan Crown steps deliberately leave Mystery null (see BeadLayout's
        // generalization) — the Seven Joys aren't Rosary "mysteries" even though 6 of the 7
        // reuse mystery imageKeys.
        var steps = BuildSteps(_engine, "en");
        Assert.All(steps, s => Assert.Null(s.Mystery));
    }

    [Fact]
    public void BuildSteps_TwoClosingHailMarysAndOneClosingOurFather()
    {
        var steps = BuildSteps(_engine, "en");
        var nonDecadeHailMarys = steps.Count(s => !s.DecadeIndex.HasValue && s.Title.StartsWith("Hail Mary"));
        Assert.Equal(2, nonDecadeHailMarys);

        var nonDecadeOurFathers = steps.Count(s => !s.DecadeIndex.HasValue && s.Title == "Our Father");
        Assert.Equal(1, nonDecadeOurFathers);
    }

    [Fact]
    public void BuildSteps_EndsWithAntiphonThenClosingCross()
    {
        var steps = BuildSteps(_engine, "en");
        Assert.True(steps[^2].IsAntiphon);
        Assert.Equal("Sign of the Cross", steps[^1].Title);
    }

    [Fact]
    public void BuildSteps_FirstJoyIsAnnunciationReusingExistingMysteryContent()
    {
        var steps = BuildSteps(_engine, "en");
        var firstJoy = steps.First(s => s.DecadeIndex == 0 && s.IsScripture);
        Assert.Equal("The Annunciation", firstJoy.Title);
        Assert.Equal("joyful_01_annunciation", firstJoy.ImageOverrideKey);
    }

    [Fact]
    public void BuildSteps_FourthJoyIsTheNewAdorationOfTheMagiContent()
    {
        var steps = BuildSteps(_engine, "en");
        var fourthJoy = steps.First(s => s.DecadeIndex == 3 && s.IsScripture);
        Assert.Equal("The Adoration of the Magi", fourthJoy.Title);
        Assert.Equal("franciscan_04_adoration_of_the_magi", fourthJoy.ImageOverrideKey);
    }

    [Fact]
    public void BuildSteps_EnglishBodyContainsEnglishText()
    {
        var steps = BuildSteps(_engine, "en");
        Assert.Contains(steps, s => s.Body.Contains("Hail Mary, full of grace"));
    }

    [Fact]
    public void BuildSteps_LatinBodyContainsLatinText()
    {
        var steps = BuildSteps(_engine, "la");
        Assert.Contains(steps, s => s.Body.Contains("Ave Maria, gratia plena"));
    }
}

File: ProsaryTests/FranciscanCrownEngineTests.swift
//
//  FranciscanCrownEngineTests.swift
//  ProsaryTests
//

import XCTest
import SwiftUI
@testable import Prosary

private struct FixedLiturgicalCalendar: LiturgicalCalendarProviding {
  var antiphon: MarianAntiphonOption = .salveRegina

  func mysteryGroup(for date: Date) -> MysteryGroup { .joyful }
  func seasonColor(for date: Date) -> Color { .clear }
  func seasonalMarianAntiphon(for date: Date) -> MarianAntiphonOption { antiphon }
  func isEasterSeason(for date: Date) -> Bool { false }
}

final class FranciscanCrownEngineTests: XCTestCase {
  private func engine(antiphon: MarianAntiphonOption = .salveRegina) -> PrayerEngine {
    PrayerEngine(calendar: FixedLiturgicalCalendar(antiphon: antiphon))
  }

  func testSevenDecadesOfTenHailMarys() {
    let steps = engine().buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "en"))
    let decadeIndices = Set(steps.compactMap(\.decadeIndex))
    XCTAssertEqual(decadeIndices, Set(0..<7))

    for d in 0..<7 {
      let hailMarysInDecade = steps.filter { $0.decadeIndex == d && $0.hailMaryIndexInDecade != nil }
      XCTAssertEqual(hailMarysInDecade.count, 10, "decade \(d) should have 10 Hail Marys")
    }
  }

  func testNoStepHasAMystery() {
    // Franciscan Crown steps deliberately leave `mystery` nil (see BeadModels' generalization) —
    // the Seven Joys aren't Rosary "mysteries" even though 6 of the 7 reuse mystery imageKeys.
    let steps = engine().buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "en"))
    XCTAssertTrue(steps.allSatisfy { $0.mystery == nil })
  }

  func testTwoClosingHailMarysAndOneClosingOurFather() {
    let steps = engine().buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "en"))
    let nonDecadeHailMarys = steps.filter { $0.decadeIndex == nil && $0.title.hasPrefix("Hail Mary") }
    XCTAssertEqual(nonDecadeHailMarys.count, 2)

    let nonDecadeOurFathers = steps.filter { $0.decadeIndex == nil && $0.title == "Our Father" }
    XCTAssertEqual(nonDecadeOurFathers.count, 1)
  }

  func testEndsWithSeasonalAntiphonThenClosingCross() {
    let steps = engine(antiphon: .reginaCaeli).buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "en"))
    XCTAssertEqual(steps[steps.count - 2].title, "Regina Caeli")
    XCTAssertTrue(steps[steps.count - 2].isAntiphon)
    XCTAssertEqual(steps.last?.title, "Sign of the Cross")
  }

  func testFirstJoyIsAnnunciationReusingExistingMysteryContent() {
    let steps = engine().buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "en"))
    let firstJoy = steps.first { $0.decadeIndex == 0 && $0.isScripture }
    XCTAssertEqual(firstJoy?.title, "The Annunciation")
    XCTAssertEqual(firstJoy?.imageKey, "joyful_01_annunciation")
  }

  func testFourthJoyIsTheNewAdorationOfTheMagiContent() {
    let steps = engine().buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "en"))
    let fourthJoy = steps.first { $0.decadeIndex == 3 && $0.isScripture }
    XCTAssertEqual(fourthJoy?.title, "The Adoration of the Magi")
    XCTAssertEqual(fourthJoy?.imageKey, "franciscan_04_adoration_of_the_magi")
  }

  func testEnglishBodyContainsEnglishText() {
    let steps = engine().buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "en"))
    XCTAssertTrue(steps.contains { $0.body.contains("Hail Mary, full of grace") })
  }

  func testLatinBodyContainsLatinText() {
    let steps = engine().buildSteps(for: Prayer(kind: .franciscanCrown, languageCode: "la"))
    XCTAssertTrue(steps.contains { $0.body.contains("Ave Maria, gratia plena") })
  }
}
